nodl_python/nodl/nodl_parse.py

```diff
@@ -0,0 +1,243 @@
+        """Parse a NoDL parameter from an xml element."""
+        return Parameter(name=element.attrib['name'], parameter_type=element.attrib['type'])
+
+    @classmethod
+    def parse_service(cls, element: etree._Element) -> Service:
+        """Parse a NoDL service from an xml element."""
+        attribs = element.attrib
+        name = attribs['name']
+        service_type = attribs['type']
+
+        server = str_to_bool(attribs.get('server', 'False'))
+        client = str_to_bool(attribs.get('client', 'False'))
+
+        if not (server or client):
+            warnings.warn(
+                f'{element.base}:{element.sourceline}: {name} is neither server or client',
+                NoNodeInterfaceWarning,
+            )
+
+        policy = element.find('qos')
+
+        return Service(
+            name=name,
+            service_type=service_type,
+            server=server,
+            client=client,
+            qos=parse_qos(policy),
+        )
+
+    @classmethod
+    def parse_topic(cls, element: etree._Element) -> Topic:
+        """Parse a NoDL topic from an xml element."""
+        attribs = element.attrib
+        name = attribs['name']
+        message_type = attribs['type']
+
+        publisher = str_to_bool(attribs.get('publisher', 'False'))
+        subscriber = str_to_bool(attribs.get('subscriber', 'False'))
+        if not (publisher or subscriber):
+            warnings.warn(
+                f'{element.base}:{element.sourceline}: {name} is neither publisher or subscriber',
+                NoNodeInterfaceWarning,
+            )
+
+        policy = element.find('qos')
+
+        return Topic(
+            name=name,
+            message_type=message_type,
+            publisher=publisher,
+            subscriber=subscriber,
+            qos=parse_qos(policy),
+        )
+
+    @classmethod
+    def parse_nodes(cls, interface) -> List[Node]:
+        """Parse the nodes contained in an interface element and return a list."""
+        node_elements = [child for child in interface if child.tag == 'node']
+        return [cls.parse_node(node) for node in node_elements]
+
+    @classmethod
+    def parse_node(cls, node: etree._Element) -> Node:
+        """Parse a NoDL node and all the elements it contains from an xml element."""
+        name = node.attrib['name']
+
+        actions = []
+        parameters = []
+        services = []
+        topics = []
+
+        for child in node:
+            try:
+                if child.tag == 'action':
+                    actions.append(cls.parse_action(child))
+                if child.tag == 'parameter':
+                    parameters.append(cls.parse_parameter(child))
+                if child.tag == 'service':
+                    services.append(cls.parse_service(child))
+                if child.tag == 'topic':
+                    topics.append(cls.parse_topic(child))
+            except KeyError as excinfo:
```

Comment:
I couldn't figure out what would be raising a KeyError for a minute, then realized that it might be coming from the individual parsing functions. I think it would be far more clear if the KeyError was caught and handled there; probably turned into an exception inheriting from our root exception and outlining exactly what the issue is. Then we don't need to handle it here.

The exception inheritance tree could look something like:

- NoDLException
  - InvalidNodeElementError
    - InvalidActionError
    - InvalidParameterError
    - ... you get the idea

Reply:
No longer relevant now that the schema handles validation?

Reply:
Agreed.